File: impl/app/android/app/src/main/java/com/ak/contexta/domain/LlmErrorClassifier.kt
package com.ak.contexta.domain

import com.ak.contexta.domain.error.PipelineBlockingException

/**
 * Classifies LLM call errors into three categories:
 * - Recoverable: network issues, 429 rate-limit, 5xx, bad JSON → retryable
 * - LlmFatal: 401/403 auth, 400 bad request, content policy → fail immediately, user can retry
 * - Structural: DB constraint violations, serialization errors → BLOCKED state
 */
object LlmErrorClassifier {

    sealed class LlmError {
        /** Recoverable — retry with backoff */
        data class Recoverable(val cause: Throwable, val retryAfterSeconds: Int? = null) : LlmError()
        /** Non-recoverable LLM-side error — fail the task, user can retry */
        data class LlmFatal(val cause: Throwable) : LlmError()
        /** Structural code-level error — block the pipeline, needs app fix */
        data class Structural(val cause: Throwable) : LlmError()
    }

    fun classify(httpCode: Int?, throwable: Throwable): LlmError {
        // Structural errors first — these are local, not LLM responses
        when {
            throwable is PipelineBlockingException -> return LlmError.Structural(throwable)
            throwable.message?.contains("Constraint") == true ->
                return LlmError.Structural(throwable)
            throwable.message?.contains("disk I/O") == true ->
                return LlmError.Structural(throwable)
        }

        val code = httpCode
        return when {
            // Recoverable
            code == null -> LlmError.Recoverable(throwable) // network error
            code == 429 -> {
                val retryAfter = extractRetryAfter(throwable.message)
                LlmError.Recoverable(throwable, retryAfter)
            }
            code in 500..599 -> LlmError.Recoverable(throwable)
            throwable.message?.contains("JSON") == true ||
                throwable.message?.contains("parse") == true ||
                throwable.message?.contains("syntax") == true -> LlmError.Recoverable(throwable)

            // LLM Fatal
            code == 400 -> LlmError.LlmFatal(throwable)
            code == 401 || code == 403 -> LlmError.LlmFatal(throwable)

            // Everything else
            else -> LlmError.Recoverable(throwable)
        }
    }

    private fun extractRetryAfter(message: String?): Int? {
        if (message == null) return null
        val match = Regex("Retry-After[=:\\s]+(\\d+)").find(message)
        return match?.groupValues?.get(1)?.toIntOrNull()
    }
}

File: impl/app/flutter/lib/domain/llm_error_classifier.dart
import 'error/pipeline_blocking_exception.dart';

/// LLM 调用错误三分类（对照 Kotlin LlmErrorClassifier.kt 的 sealed class）。
///
/// - [RecoverableError] — 网络问题、429 限流、5xx、坏 JSON → 可重试
/// - [FatalError] — 401/403 鉴权、400 坏请求、内容策略 → 立即失败，用户可重试
/// - [StructuralError] — DB 约束冲突、序列化错误 → BLOCKED 状态
sealed class LlmError {
  const LlmError(this.cause);

  /// 原始异常（Kotlin: cause）。
  final Object cause;
}

/// 可恢复 — 退避重试。
class RecoverableError extends LlmError {
  const RecoverableError(super.cause, {this.retryAfterSeconds});

  /// 429 时服务端建议的等待秒数（Retry-After 头）。
  final int? retryAfterSeconds;
}

/// LLM 侧不可恢复错误 — 立即失败任务，用户可重试。
class FatalError extends LlmError {
  const FatalError(super.cause);
}

/// 代码级结构性错误 — 阻塞流水线，需应用修复。
class StructuralError extends LlmError {
  const StructuralError(super.cause);
}

/// 分类器（纯函数，对照 Kotlin object LlmErrorClassifier）。
class LlmErrorClassifier {
  LlmErrorClassifier._();

  static LlmError classify(int? httpCode, Object error) {
    // 结构性错误优先——它们是本地错误，不是 LLM 响应
    final message = error is Exception ? error.toString() : error.toString();
    if (error is PipelineBlockingException) {
      return StructuralError(error);
    }
    if (message.contains('Constraint')) {
      return StructuralError(error);
    }
    if (message.contains('disk I/O')) {
      return StructuralError(error);
    }

    final code = httpCode;
    if (code == null) {
      // 网络错误（未解析出 HTTP 状态码）→ 可恢复
      return RecoverableError(error);
    }
    if (code == 429) {
      return RecoverableError(error, retryAfterSeconds: _extractRetryAfter(message));
    }
    if (code >= 500 && code <= 599) {
      return RecoverableError(error);
    }
    if (message.contains('JSON') ||
        message.contains('parse') ||
        message.contains('syntax')) {
      return RecoverableError(error);
    }

    // LLM 致命错误
    if (code == 400 || code == 401 || code == 403) {
      return FatalError(error);
    }

    // 其他 → 保守可恢复
    return RecoverableError(error);
  }

  /// 从异常消息提取 Retry-After 秒数（Kotlin 正则 `Retry-After[=:\s]+(\d+)`）。
  static int? _extractRetryAfter(String message) {
    final match = RegExp(r'Retry-After[=:\s]+(\d+)').firstMatch(message);
    if (match == null) return null;
    return int.tryParse(match.group(1)!);
  }
}
